airflow/models/dagrun.py

```diff
@@ -631,6 +631,68 @@ def update_state(
         session.merge(self)
         # We do not flush here for performance reasons(It increases queries count by +20)
 
+        from airflow.models import Dataset
```

Comment:
For clarity (and perhaps ease of testing) I would suggest moving all this logic to a separate (private?) function

Reply:
done